src/SurrogatesBase.jl

```diff
@@ -0,0 +1,141 @@
+module SurrogatesBase
+
+import Statistics: mean, var
+import Base: rand
+import StatsBase: mean_and_var
+
+export AbstractSurrogate, add_point!
+export update_hyperparameters!, hyperparameters
+export mean, var, mean_and_var, rand
+
+"""
+    abstract type AbstractSurrogate{D, R} end
+
+An abstract type for surrogates, parametrized by domain type `D` and
+range type `R` of the underlying function that is being approximated.
+
+    (s::AbstractSurrogate{D})(x::D) where D
+
+Subtypes of `AbstractSurrogate` can be callable with input points `x` such that the result
+is an evaluation of the surrogate at `x`, corresponding to an approximation of the underlying
+function at `x`.
+
+ # Examples
+ ```jldoctest
+ julia> struct ZeroSurrogate{D, R} <: AbstractSurrogate{D, R} end
+
+ julia> (::ZeroSurrogate{D})(x::D) where D = 0
+
+ julia> s = ZeroSurrogate{Int, Int}()
+ ZeroSurrogate{Int64, Int64}()
+
+ julia> s(4) == 0
+ true
+ ```
+ """
+abstract type AbstractSurrogate{D, R} <: Function end
+
+"""
+    add_point!(s::AbstractSurrogate{D, R}, new_x::D, new_y::R) where {D, R}
+
+Add an evaluation `new_y` at point `new_x` to the surrogate.
+"""
+function add_point! end
+"""
+    add_point!(s::AbstractSurrogate{D, R}, new_xs::AbstractVector{D}, new_ys::AbstractVector{R}) where {D, R}
+
+Add evaluations `new_ys` at points `new_xs` to the surrogate.
+
+Use `add_point!(s, eachslice(X, dims = 2), new_ys)` if `X` is a matrix.
+"""
+function add_point!(s::AbstractSurrogate{D, R},
+    new_xs::AbstractVector{D},
+    new_ys::AbstractVector{R}) where {D, R}
+    add_point!.(Ref(s), new_xs, new_ys)
+end
+
+"""
+    update_hyperparameters!(s::AbstractSurrogate, prior)
+
+Use prior on hyperparameters passed in `prior` to perform an update.
+
+See also [`hyperparameters`](@ref).
+"""
+function update_hyperparameters! end
+
+"""
+    hyperparameters(s::AbstractSurrogate)
+
+Return a `NamedTuple`, in which names are hyperparameters and values are currently used
+values of hyperparameters in `s`.
+
+See also [`update_hyperparameters!`](@ref).
+"""
+function hyperparameters end
+
+"""
+    mean(s::AbstractSurrogate{D}, x::D) where D
+
+Return mean at point `x`.
+"""
+function mean end
```

Comment:
This is benign type piracy but still type piracy and something we should avoid

```suggestion
```

Reply:
Sorry, I don't quite understand, why is this considered a type piracy? We are extending the documentation of `mean` for *our own* datatype used in the argument. Like if we define `mean(s::AbstractSurrogate{D}, x::D) where D = error("..")` and add the same docs as above, it would not be considered a piracy, right?

Reply:
AFAIK you can fix the type piracy with
```suggestion
"""
    mean(s::AbstractSurrogate{D}, x::D) where D

Return mean at point `x`.
"""
mean(::AbstractSurrogate{D}, x::D) where {D}
```

But generally my impression is that it is a misuse of and inconsistent with the definition of `Statistics.mean`: the surrogate `s` is a function of `x`, but for such a constellation `mean` is already defined in a different way in Statistics as shown in the docstring:
```julia
mean(f, itr)

  Apply the function f to each element of collection itr and take the mean.

  julia> using Statistics

  julia> mean(√, [1, 2, 3])
  1.3820881233139908

  julia> mean([√1, √2, √3])
  1.3820881233139908
```

Reply:
You are right, we would be extending the function for a different purpose, thanks for pointing this out!